design/20220722-publicbundle.md

```diff
@@ -0,0 +1,289 @@
+	"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem", // CentOS/RHEL 7
+	"/etc/ssl/cert.pem",                                 // Alpine Linux
+}
+```
+
+It's a chore to manually specify a location for a trust store, and it's unfamiliar and error-prone
+for many users. However, if a user instead overwrites any of these files with their own
+`trust`-generated certificate trust bundle they'll currently overwrite their public trust bundle
+provided by their OS / container base image.
+
+Unless the user's application only talks to services which the user also controls, what users are
+likely to really want in most cases is a bundle containing both the publicly trusted certs _and_
+the user's own roots.
+
+It's possible in most distros to append to the cert bundle, but that usually involves running some
+command to generate the final bundle, which would change how the user invokes their application
+inside their container.
+
+Ideally, the whole bundle a user requires would simply be available at one of the above locations.
+
+### Goals
+
+- A user configuring a `Bundle` can add a public trust bundle at least as easily as they can add certs from a `ConfigMap` or `Secret`
+- It's clear to users exactly what version of the public bundle is used at any given time, for auditing purposes
+- A user whose application must use both public and private CAs can confidently overwrite their "system" trust store in a container using a trust `Bundle`
+- A user can update any bundle in use with any version of the trust controller and be confident that it will work
+
+### Non-Goals
+
+- Editing/pruning the public trust bundle (out of scope for the initial feature but might be desirable later)
+- Supporting [OS-specific](https://superuser.com/questions/437330/how-do-you-add-a-certificate-authority-ca-to-ubuntu) certificate update mechanisms
+- Supporting a plethora of options for sources of public trust bundles; we might want this in the future but we should start with one
+
+## Proposal
+
+We propose that the cert-manager project becomes a distributor of its own public trust bundle, which at
+least to start will be a carbon copy of the Mozilla trust bundle.
+
+That bundle in will in turn be packaged into a container which will be used as a sidecar in the trust
+controller's pod, with the ability to present its own bundle to the controller container.
+
+A deployed instance of trust will therefore contain two containers in a pod; the trust controller itself, and a bundle
+container with which trust will communicate in order to fetch a trust bundle.
+
+We'll also add a `publicBundle` field as a source, which is the method by which users include the bundle from the bundle container.
+
+```yaml
+apiVersion: trust.cert-manager.io/v1alpha1
+kind: Bundle
+metadata:
+  name: my-org.com
+spec:
+  sources:
+  - configMap:
+      name: "my-root-cm"
+      key: "root-cert.pem"
+  - publicBundle:
+      source: "cert-manager" # i.e. use the cert-manager bundle
```

Comment:
I'd rather we be more specific, and use `mozilla` or `cert-manager/mozilla`.

Reply:
Yeah I'm easy about the name really, it's not important!